save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt

```diff
@@ -0,0 +1,262 @@
+                .map { it.toDto() }
+        }
+
+    /**
+     * @param organizationName
+     * @param name
+     * @return [TestSuitesSourceDto] found by provided values or not found exception
+     */
+    @GetMapping("/{organizationName}/{name}")
+    fun findAsDtoByName(
+        @PathVariable organizationName: String,
+        @PathVariable name: String
+    ): Mono<TestSuitesSourceDto> = Mono.just(organizationName)
+        .flatMap {
+            organizationService.findByName(it).toMono()
+        }
+        .switchIfEmptyToResponseException(HttpStatus.CONFLICT) {
+            "Organization not found by name $organizationName"
+        }
+        .flatMap { organization ->
+            testSuitesSourceService.findByName(organization, name).toMono()
+        }
+        .switchIfEmptyToNotFound {
+            "TestSuitesSource not found by name $name for organization $organizationName"
+        }
+        .map { it.toDto() }
+
+    /**
+     * Upload snapshot of [TestSuitesSource] with [version]
+     *
+     * @param organizationName
+     * @param name
+     * @param version
+     * @param creationTime
+     * @param contentAsMonoPart
+     * @return empty response
+     */
+    @PostMapping("/{organizationName}/{name}/upload-snapshot", consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
+    fun uploadSnapshot(
+        @PathVariable organizationName: String,
+        @PathVariable name: String,
+        @RequestParam version: String,
+        @RequestParam creationTime: Long,
+        @RequestPart("content") contentAsMonoPart: Mono<Part>
+    ): Mono<Unit> = findAsDtoByName(organizationName, name)
+        .map { TestSuitesSourceSnapshotKey(it, version, creationTime) }
+        .flatMap { key ->
+            contentAsMonoPart.flatMap { part ->
+                val content = part.content().map { it.asByteBuffer() }
+                testSuitesSourceSnapshotStorage.upload(key, content).map { writtenBytes ->
+                    log.info { "Saved ($writtenBytes bytes) snapshot of ${key.testSuitesSourceName} in ${key.organizationName} with version $version" }
+                }
+            }
+        }
+
+    /**
+     * Download snapshot of [TestSuitesSource] with [version]
+     *
+     * @param organizationName
+     * @param name
+     * @param version
+     * @return resource response
+     */
+    @PostMapping("/{organizationName}/{name}/download-snapshot", produces = [MediaType.APPLICATION_OCTET_STREAM_VALUE])
+    fun downloadSnapshot(
+        @PathVariable organizationName: String,
+        @PathVariable name: String,
+        @RequestParam version: String,
+    ): Mono<ByteBufferFluxResponse> = testSuitesSourceSnapshotStorage.findKey(organizationName, name, version)
+        .map {
+            ResponseEntity.ok()
+                .contentType(MediaType.APPLICATION_OCTET_STREAM)
+                .body(testSuitesSourceSnapshotStorage.download(it))
+        }
+        .onErrorReturn(
```

Comment:
`switchIfEmptyToNotFound` is not applicable here?

Reply:
will add

Reply:
added for case when key is not found